### PR TITLE
[scheduler] Add a prop to control whether the resource of an event can be cleared

### LINKED ISSUE #20587: [scheduler] Add a prop to control whether the resource of an event can be cleared
Based on https://github.com/mui/mui-x/pull/20553#discussion_r2592703744

Both Event Calendar and Event Timeline should expose a prop that controls whether the resource of an event can be cleared in the edit event dialog. The prop is the same on both components; only the default changes:

| Component | Default | Reason |
  |---|---|---|
  | Event Timeline | Resource required (can't be cleared) | Events are grouped by resource — clearing it makes the event disappear from the view |
  | Event Calendar | Resource optional (can be cleared) | No visual dependency on the resource |

Why keep it configurable on Timeline: in the future we may add a dedicated row for events with no resource, at that point users will want to opt in and allow clearing the resource on Timeline too (https://github.com/mui/mui-x/issues/22149)

Naming: we plan to support multiple resources per event in the future, so the prop name should be plural-friendly from day one (e.g. `requireResources`) rather than singular.

### PR DESCRIPTION
**Pull request by rita-codes**
Closes #20587.

Adds a shared `shouldEventRequireResource` prop on `EventCalendar`, `EventCalendarPremium`, and `EventTimelinePremium` to control whether an event's resource can be cleared in the edit dialog.

  - Defaults: `false` on Calendar, `true` on Timeline (where events without a resource aren't rendered).
  - When `true`, the "No resource" option is hidden in the resource picker and submitting an event with no resource is blocked.

- [x] I have followed (at least) the [PR section of the contributing guide](https://github.com/mui/mui-x/blob/HEAD/CONTRIBUTING.md#sending-a-pull-request).

### CODE REVIEW

**Review by flaviendelangle — COMMENTED** [review]
Maybe `allowEventWithoutResource` would be more explicit?
Or `shouldEventHaveResource`

**Review by flaviendelangle — COMMENTED** [review]
# PR #22464 Review — `requireResources` prop

**Scope:** 79 files, +398 / −13. Adds `requireResources` to EventCalendar (default `false`), EventCalendarPremium (`false`), and EventTimelinePremium (`true`). When true, hides the "No resource" option and blocks submission of events without a resource.

---

## Critical (fix before merge)

1. **Pre-existing events without a resource become un-editable on Timeline** *(silent-failure)*
   Timeline ships with `requireResources={true}`. Any consumer with existing events that have `resource: undefined` can open the dialog, sees the misleading `labelInvalidResource` text in the Select, and cannot save *any* unrelated edit (title, description, dates) without assigning a resource. No migration path is documented, no warning fires. — [ResourceAndColorSection.tsx:148-169](packages/x-scheduler/src/internals/components/event-dialog/ResourceAndColorSection.tsx#L148-L169)

2. **Submit silently no-ops — no visible feedback** *(silent-failure)*
   The Save button has no `disabled` / `aria-disabled` / `aria-describedby`. On first click in the no-resource state the form does nothing visible until the user scrolls to the resource field. Recommend disabling the button (with tooltip) or scrolling/focusing the field + `role="alert"` on the helper text. — [FormContent.tsx:309-311](packages/x-scheduler/src/internals/components/event-dialog/FormContent.tsx#L309-L311)

3. **JSDoc text is inaccurate** *(comment-analyzer)*
   "Whether each event must have at least one resource assigned." — events have at most *one* resource (single `resourceId`; guard is `controlled.resourceId === null`). Replicated across 1 type def + 7 propTypes + 7 translation JSONs + 7 API JSONs. Suggested: *"Whether each event must be assigned to a resource. When `true`, the resource cannot be cleared in the edit dialog and the form cannot be submitted without one."* — [SchedulerStore.types.ts:255](packages/x-scheduler-internals/src/internals/utils/SchedulerStore/SchedulerStore.types.ts#L255)

4. **Race condition: `requireResources` toggled while dialog is open** *(silent-failure)*
   Flipping the prop while the dialog is open with `resourceId === null` makes the selected option disappear from the menu but `controlled.resourceId` stays null. The Select then shows "Invalid resource". `errors` local state isn't reconciled when the prop changes either. — [EventCalendarStore.ts](packages/x-scheduler-internals/src/use-event-calendar/EventCalendarStore.ts), [FormContent.tsx:133](packages/x-scheduler/src/internals/components/event-dialog/FormContent.tsx#L133)

5. **Test gap: no `updater` test for `requireResources`** *(pr-test-analyzer)*
   Only initial state is asserted in both store tests. A regression removing the line from `updateStateFromParameters` would be invisible. Add an updater test mirroring the existing `views` updater pattern. — [core.EventCalendarStore.test.ts:58-69](packages/x-scheduler-internals/src/use-event-calendar/tests/core.EventCalendarStore.test.ts#L58-L69), [core.EventTimelinePremiumStore.test.ts:51-65](packages/x-scheduler-internals-premium/src/use-event-timeline-premium/tests/core.EventTimelinePremiumStore.test.ts#L51-L65)

---

## Important

6. **MUI Select "out-of-range value" dev warning** *(code-reviewer)*
   With `requireResources={true}` and `resourceId == null`, the Select receives `value=""` but the `NO_RESOURCE_VALUE` MenuItem is no longer rendered → MUI logs *"You have provided an out-of-range value…"*. Triggers on every new event in Timeline (the default config). Keep the empty option in the list but hide it with `display: none` when `requireResources`. — [ResourceAndColorSection.tsx:219](packages/x-scheduler/src/internals/components/event-dialog/ResourceAndColorSection.tsx#L219)

7. **Error not cleared when user picks a resource after a failed submit** *(code-reviewer + silent-failure)*
   `handleResourceChange` does not call `setErrors(...)` while other change handlers do. Helper text remains visible until next Save click. — [GeneralTab.tsx:145-149](packages/x-scheduler/src/internals/components/event-dialog/GeneralTab.tsx#L145-L149)

8. **Per-component API doc default is confusing** *(code-reviewer + comment-analyzer)*
   `@default false on EventCalendar / EventCalendarPremium, true on EventTimelinePremium` is rendered verbatim on *every* component's API page. Either override `@default` at each public-component prop declaration so each page reads the appropriate boolean, or shorten the centralized default to point at the component docs. — [SchedulerStore.types.ts:254](packages/x-scheduler-internals/src/internals/utils/SchedulerStore/SchedulerStore.types.ts#L254)

9. **Form helper text + Select are not wired for screen readers** *(silent-failure)*
   `<FormHelperText>` has no `id`; the `<Select>` has no `aria-describedby` / `aria-errormessage`. SR users tabbing back hear nothing about the error. — [ResourceAndColorSection.tsx:886](packages/x-scheduler/src/internals/components/event-dialog/ResourceAndColorSection.tsx#L886)

10. **`warnOnce` for "no resources configured" only fires while the dialog is mounted** *(silent-failure)*
    Move the misconfiguration check to the store's `updateStateFromParameters` / mapper so it surfaces at mount time regardless of whether the user ever opens the dialog. — [ResourceAndColorSection.tsx:826-834](packages/x-scheduler/src/internals/components/event-dialog/ResourceAndColorSection.tsx#L826-L834)

11. **Enforcement is dialog-only — drift with `createEvent` / `updateEvent`** *(silent-failure)*
    The store mutations still accept `resource: undefined`. Drag-to-create or any non-dialog code path can produce resource-less events while `requireResources={true}`. Either rename the prop to be honest about scope (`requireResourceInDialog`) or enforce it in the store mutations with a logged error. — [FormContent.tsx:193,214,223](packages/x-scheduler/src/internals/components/event-dialog/FormContent.tsx#L193)

12. **Generic error string violates project guidelines** *(silent-failure)*
    `'A resource is required.'` doesn't follow AGENTS.md "say what / why / how to fix". Suggested: *"Select a resource for this event before saving."* — [enUS.ts](packages/x-scheduler/src/locales/enUS.ts)

13. **Test gap: no public-API smoke test for the 6 entry components** *(pr-test-analyzer)*
    `EventCalendarPremium`, `EventTimelinePremium`, and the four `Standalone*View` components add the prop but no test asserts it actually reaches the dialog via the `useExtract*Parameters` destructure. A future refactor that forgets the destructure is invisible.

14. **Test gap: `warnOnce` dev warning untested** *(pr-test-analyzer)*
    Add a `toWarnDev` test rendering with `requireResources resources={[]}`.

15. **Test gap: submit-error condition relies on implementation detail** *(pr-test-analyzer)*
    Guard is `=== null`. If the controller is ever refactored to leave `undefined`, the guard silently passes. Add a test for `resource: undefined` mapping to `null` (or document the contract).

---

## Suggestions

16. Non-EN locales (frFR, deDE, itIT, esES, ptBR, ptPT, roRO are complete) leave `requiredResourceError` commented out — add translations before merge. — [packages/x-scheduler/src/locales/](packages/x-scheduler/src/locales/)
17. JSDoc never mentions the two observable behaviors (hides option, blocks submit) — expanding to two sentences makes IDE tooltips self-sufficient.
18. Redundant `process.env.NODE_ENV !== 'production'` wrapper around `warnOnce` (the helper already short-circuits in production). — [ResourceAndColorSection.tsx:147-155](packages/x-scheduler/src/internals/components/event-dialog/ResourceAndColorSection.tsx#L147-L155)
19. DRY the four `requireResources` dialog tests — they share fixtures.
20. Add a test that the helper text uses the **locale** key, not hard-coded English (catches missing locale entries falling back silently).
21. `labelInvalidResource` is shown for a *missing* (not invalid) resource — misleading copy; consider a separate `labelMissingResource` key.

---

## Strengths

- Both defaults correctly differentiated and explicitly asserted in store tests.
- `Omit<SchedulerState<TEvent>, 'requireResources'>` cleanly forces each mapper to supply the state — good use of types.
- Prop threaded correctly through both `useExtract*Parameters` hooks (both `parameters` and `forwardedProps` spreads).
- Locale type + all 41 locale files updated in lockstep.
- API JSON + translation JSON regenerated via the scripts (commits `proptypes` / `docs api`), not hand-edited.
- JSDoc on `getInitialState` explaining *why* `requireResources` is `Omit`ted is well written.
- `warnOnce` message itself follows project conventions (MUI X prefix, what/why/how).
- Submit-blocked test is behavioral (clicks Save, asserts no callback + visible error), not flag-based.

---

## Recommended action plan

1. Fix Critical #1 (existing-event trap) — likely needs either a "keep current resourceless events visible in the picker" branch, or explicit docs that the prop only applies to *new* events.
2. Fix Critical #2 (no submit feedback) and Important #9 (a11y wiring) — these go together.
3. Fix Critical #3 (JSDoc text) — single coordinated edit across type + 7 components + regenerate JSONs.
4. Fix Critical #4 (race) — clear `errors.resource` on `requireResources` change.
5. Fix Critical #5 + Important #13–15 (test gaps).
6. Address Important #6–#12.
7. Re-run review on the updated diff.

**Comment by rita-codes** [comment]
> Maybe `allowEventWithoutResource` would be more explicit? Or `shouldEventHaveResource`

Good point on the ambiguity @flaviendelangle , `requireResources` mixes two things (resources existing vs events being assigned to one). I'd lean toward `eventResourcesRequired`, which makes the assignment rule explicit while keeping it positive and plural-friendly (the ticket calls for plural since multi-resource per event is on the roadmap, and the rule is "at least one resource per event")

On the other directions:
- `allowEventWithoutResource(s)`: reads as a double negative when false, so I'd rather keep it positive.
- `shouldEventHaveResources`: should reads as a recommendation), but this is hard validation 👀 

Other names that fit positive + plural + assignment-explicit if `resourcesRequired` feels off: `requireEventResources`, `enforceEventResources`, `eventResourcesMandatory`. Preference or any other idea?

**Comment by rita-codes** [comment]
Thanks for the review @flaviendelangle !
Here the details for the fixes I applied:
1. Good catch, the `labelInvalidResource` fallback is misleading here: an unassigned resource isn't the same as an invalid one. Today the case is only reachable programmatically (Timeline doesn't render events without a resource), but #22149 will make it user-reachable. I'll add a distinct empty-state for `resourceId == null` when `requireResources={true}`. The "silent failure" framing still doesn't quite fit IMO, `requiredResourceError` is shown on save, but the UX point stands.
2. Pushback on the "silent-failure" framing: the form is short and the resource field sits right above the description and Save button, the error helper is visible without scrolling in normal viewports. I wouldn't disable the Save button either (a11y anti-pattern: disabled buttons don't receive focus and can't announce why they're disabled to screen readers). That said, you're right that screen readers don't get an announcement when the error appears after clicking Save, `TextField` only wires `aria-describedby`, which is read on-focus, not on-appear. Added `role="alert”` to the helper text of all three error fields in the dialog (`title`, `startDate`, `resource`) so the error gets announced live regardless of focus
3. Fair point, “at least one" reads off when events only have a single resource today. Reworded.
4. The "Invalid resource" outcome no longer applies — the Select fix in this PR distinguishes unassigned (null) from invalid and shows labelNoResource for the former. As for the race: requireResources is a configuration prop, not meant to be toggled dynamically at runtime; the consuming app sets it once at mount. handleSubmit revalidates against the current prop value, so no invalid save sneaks through — there's no silent failure. The stale errors state clears on the next interaction. I'd leave this as-is.
5. Tests added.
6. Fixed
7. Fixed
8. Fixed
9. Fixed
10. Fixed
11. The prop is intentionally scoped to the edit dialog per the ticket — the feature is about whether the user can clear the resource via the UI, not a data-layer invariant. Drag-to-create on Timeline already produces events with a resource (the row the user dropped on). Programmatic createEvent / updateEvent calls are user-land — if the consumer wants strict validation there, they can wrap their own. Renaming to requireResourceInDialog would break the plural future-proofing the ticket calls out, and adding store-level enforcement extends the feature beyond its specified scope. Happy to revisit if there's a concrete use case where this surfaces.
12. The error message guidelines apply to errors thrown from public packages,  `requiredResourceError` is a UI form validation message rendered under the field, not a thrown error, and it follows the same concise convention as the other form errors in the same file.
13. Fixed
14. Fixed
15. The contract you're worried about is already exercised end-to-end by the existing test should block submit and not call `onEventsChange` when `requireResources={true}` and the event has no resource in `EventDialog.test.tsx`.
16. Spanish added.
17. Already updated.
18. Fixed
19. Fixed
20. The locale system itself has dedicated infrastructure tests
21. Fixed with previous fixes.
22. 
\+ rename `requireResources`, now is `shouldEventRequireResource`

File: docs/data/scheduler/event-calendar/resources/resources.md [hunk 1]
```diff
@@ -80,6 +80,15 @@ return (
 
 :::
 
+## Require a resource
+
+By default, an event on the Event Calendar can be saved without a resource — the edit dialog includes a "No resource" option in the resource picker.
+Set `requireResources` to `true` to make the resource mandatory: the "No resource" option is hidden and the form cannot be submitted with an empty resource.
+
+```tsx
+<EventCalendar requireResources />
```
Comment (noraleonte) [inline docs/data/scheduler/event-calendar/resources/resources.md]: 🙈 
```suggestion
<EventCalendar shouldEventRequireResource />
```

**Review by noraleonte — COMMENTED** [review]
# PR #22464 Review — `shouldEventRequireResource`

**Scope:** 114 files, +952 / −206. Adds a prop (originally `requireResources`, renamed during review) to make the resource selection mandatory in the event dialog. Defaults: `false` on `EventCalendar` / `EventCalendarPremium`, `true` on `EventTimelinePremium`.

## Critical

1. **Docs still use the old prop name `requireResources`** — both demo markdowns at `docs/data/scheduler/event-calendar/resources/resources.md` (lines 5–16) and `docs/data/scheduler/event-timeline/resources/resources.md` (lines 21–39) still show:
   ```tsx
   <EventCalendar requireResources />
   <EventTimelinePremium requireResources={false} />
   ```
   The rename to `shouldEventRequireResource` was not propagated to docs. Copy-pasting these examples will silently do nothing.

## Important

2. **`setErrors({})` is too aggressive** at `GeneralTab.tsx:145`. Picking a resource now wipes title and start-date errors as well, which the user hasn't fixed. Should be `setErrors((prev) => ({ ...prev, resource: undefined }))`.

3. **Runtime-toggle behavior is implemented but unreconciled** — `updateStateFromParameters` does sync the prop (`EventCalendarStore.ts` ~lines 2100–2108), and there's a test asserting it ("should sync `shouldEventRequireResource` when parameters update"). But the dialog's `errors` local state is not reconciled, so flipping `true → false` while the dialog is open leaves a stale "A resource is required" message until the next interaction. Either drop runtime-sync support or clear `errors.resource` when the flag changes. The "we don't support runtime toggling" defense conflicts with the code that supports runtime toggling.

## Minor

4. **`aria-hidden` on focusable MenuItem** at `ResourceAndColorSection.tsx:269`. `display:none` already removes it from the a11y tree and keyboard nav; the explicit `aria-hidden` on top is redundant and is the kind of pattern axe flags. Pick one.

5. **Test fixture coupling** — ~30 test files had to be updated to pass `resources: [...]` or `shouldEventRequireResource: false` to silence the new dev warning when constructing the store. That's a lot of collateral damage for an opt-in feature. Consider gating the warning behind first dialog open (or first save) so existing test setups don't have to know about resources.

6. **`hidden?: boolean` on `ResourceOptionType`** is set but never read for any purpose other than `aria-hidden` + `display:none` styling. It's a one-shot field — could be inlined as a boolean expression at the JSX site, but keeping it on the option keeps the memo clean. Minor.

7. **Comment at `ResourceAndColorSection.tsx:155-158`** explains a non-obvious workaround (MUI out-of-range warning) — good. Per project guidance, this is the right kind of comment.

## Recommended actions before merge

1. **Must:** Fix the prop name in both `docs/data/scheduler/*/resources/resources.md` files.
2. **Should:** Scope `setErrors` to clear only the `resource` key (finding #2).
3. **Should:** Decide on runtime-toggle semantics — either reconcile `errors.resource` or remove the updater path + the corresponding test (finding #3).
4. **Nice:** Drop the redundant `aria-hidden` on the hidden MenuItem (finding #4).

**Comment by rita-codes** [comment]
> # PR #22464 Review — `shouldEventRequireResource`
> **Scope:** 114 files, +952 / −206. Adds a prop (originally `requireResources`, renamed during review) to make the resource selection mandatory in the event dialog. Defaults: `false` on `EventCalendar` / `EventCalendarPremium`, `true` on `EventTimelinePremium`.
> 
> ## Critical
> 1. **Docs still use the old prop name `requireResources`** — both demo markdowns at `docs/data/scheduler/event-calendar/resources/resources.md` (lines 5–16) and `docs/data/scheduler/event-timeline/resources/resources.md` (lines 21–39) still show:
>    ```tsx
>    <EventCalendar requireResources />
>    <EventTimelinePremium requireResources={false} />
>    ```
>    
>    
>        
>          
>        
>    
>          
>        
>    
>        
>      
>    The rename to `shouldEventRequireResource` was not propagated to docs. Copy-pasting these examples will silently do nothing.
> 
> ## Important
> 2. **`setErrors({})` is too aggressive** at `GeneralTab.tsx:145`. Picking a resource now wipes title and start-date errors as well, which the user hasn't fixed. Should be `setErrors((prev) => ({ ...prev, resource: undefined }))`.
> 3. **Runtime-toggle behavior is implemented but unreconciled** — `updateStateFromParameters` does sync the prop (`EventCalendarStore.ts` ~lines 2100–2108), and there's a test asserting it ("should sync `shouldEventRequireResource` when parameters update"). But the dialog's `errors` local state is not reconciled, so flipping `true → false` while the dialog is open leaves a stale "A resource is required" message until the next interaction. Either drop runtime-sync support or clear `errors.resource` when the flag changes. The "we don't support runtime toggling" defense conflicts with the code that supports runtime toggling.
> 
> ## Minor
> 4. **`aria-hidden` on focusable MenuItem** at `ResourceAndColorSection.tsx:269`. `display:none` already removes it from the a11y tree and keyboard nav; the explicit `aria-hidden` on top is redundant and is the kind of pattern axe flags. Pick one.
> 5. **Test fixture coupling** — ~30 test files had to be updated to pass `resources: [...]` or `shouldEventRequireResource: false` to silence the new dev warning when constructing the store. That's a lot of collateral damage for an opt-in feature. Consider gating the warning behind first dialog open (or first save) so existing test setups don't have to know about resources.
> 6. **`hidden?: boolean` on `ResourceOptionType`** is set but never read for any purpose other than `aria-hidden` + `display:none` styling. It's a one-shot field — could be inlined as a boolean expression at the JSX site, but keeping it on the option keeps the memo clean. Minor.
> 7. **Comment at `ResourceAndColorSection.tsx:155-158`** explains a non-obvious workaround (MUI out-of-range warning) — good. Per project guidance, this is the right kind of comment.
> 
> ## Recommended actions before merge
> 1. **Must:** Fix the prop name in both `docs/data/scheduler/*/resources/resources.md` files.
> 2. **Should:** Scope `setErrors` to clear only the `resource` key (finding [Color Accessibility threshold #2](https://github.com/mui/mui-x/issues/2)).
> 3. **Should:** Decide on runtime-toggle semantics — either reconcile `errors.resource` or remove the updater path + the corresponding test (finding [Cells typography #3](https://github.com/mui/mui-x/issues/3)).
> 4. **Nice:** Drop the redundant `aria-hidden` on the hidden MenuItem (finding [Surface code review  #4](https://github.com/mui/mui-x/issues/4)).

Thanks for the review @noraleonte !!!

Here the details of the fixes:
-1. Fixed
-2. Fixed
-3. Fixed.
-4. Fixed.
-5.   The warning used to live in ResourceAndColorSection and only fired on first dialog open, I intentionally moved it to the store mapper because the misconfig is a soft-lock (no event can ever be saved) and we wanted it surfaced before the UI is exercised. The fixture updates also bring the test setups closer to realistic usage: a real app with  shouldEventRequireResource: true would always pair it with resources, so tests that didn't were silently in an invalid state. Deferring the warning would re-introduce the gap where the misconfig only shows up if someone opens the dialog in dev.
-6. Leaving hidden on the option — as you noted, it keeps the memo as the single source for "should this row render," and the JSX stays free of sentinel checks like value === null. Happy to inline if you'd rather, but I prefer this shape.
-7. :D

**Review by noraleonte — APPROVED** [review]
Thanks for your work on this one! 🙏 
LGTM 👌 🎉
